File: src/boxing_coach/analysis/rules/hands_up.py
"""Rule: are the hands kept up when *not* punching?

Distinct from guard return (which is about the moment after a punch). This is
the baseline habit: between exchanges, do the wrists stay up near the head, or
do they drift down below the shoulder line? We measure the fraction of idle
frames spent with hands down.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from ...domain.analysis import Observation, Severity, SkillCategory
from ...domain.landmarks import Landmark, Side
from .. import geometry as geo
from ..context import AnalysisContext
from ..rule import Rule


@dataclass(frozen=True, slots=True)
class HandsUpConfig:
    # A hand is "down" when the wrist sits this far below the shoulder line
    # (in torso-lengths). Small buffer avoids flagging a hand right at the line.
    drop_margin: float = 0.10
    # Flag the round if a hand is down for more than this fraction of idle time.
    max_down_fraction: float = 0.25
    # Which hands to judge. A Philly shell keeps its lead hand deliberately low,
    # so that style checks the rear hand only (check_lead=False).
    check_lead: bool = True
    check_rear: bool = True


class HandsUpRule(Rule):
    id = "hands_up"
    focus_tags = frozenset()  # always relevant

    def __init__(self, config: HandsUpConfig | None = None) -> None:
        self._cfg = config or HandsUpConfig()

    def evaluate(self, context: AnalysisContext) -> list[Observation]:
        cfg = context.style_profile.config_for(self.id, self._cfg)
        idle = self._idle_frame_indices(context)
        if not idle:
            return []

        stance = context.drill.stance
        checked: set[Side] = set()
        if cfg.check_lead:
            checked.add(stance.lead)
        if cfg.check_rear:
            checked.add(stance.rear)

        observations: list[Observation] = []
        for side in (Side.LEFT, Side.RIGHT):
            if side not in checked:
                continue
            down_fraction, worst_ms = self._down_fraction(context, side, idle, cfg)
            if down_fraction is None:
                continue
            if down_fraction > cfg.max_down_fraction:
                hand = "lead" if side is stance.lead else "rear"
                observations.append(
                    Observation(
                        rule_id=self.id,
                        category=SkillCategory.DEFENCE,
                        severity=Severity.MODERATE if down_fraction > 0.5 else Severity.MINOR,
                        coaching_text=(
                            f"Your {hand} hand keeps drifting down between punches. "
                            f"Glue it to your cheek — you're open to the counter."
                        ),
                        timestamp_ms=worst_ms,
                        metrics={"down_fraction": round(down_fraction, 3)},
                        highlight_landmarks=(side.wrist,),
                    )
                )
        return observations

    def _idle_frame_indices(self, context: AnalysisContext) -> list[int]:
        """Frames that aren't part of any punch (start..end inclusive)."""
        punching = set()
        for p in context.punches:
            punching.update(range(p.start_index, p.end_index + 1))
        return [i for i in range(len(context.sequence)) if i not in punching]

    def _down_fraction(
        self, context: AnalysisContext, side: Side, idle: list[int], cfg: HandsUpConfig
    ) -> tuple[float | None, float | None]:
        seq = context.sequence
        scale = context.body_scale
        considered = 0
        down = 0
        worst_ms: float | None = None
        worst_drop = 0.0
        for i in idle:
            frame = seq.frames[i]
            wrist = geo.frame_point(frame, side.wrist)
            shoulder = geo.frame_point(frame, side.shoulder)
            if np.any(np.isnan(wrist)) or np.any(np.isnan(shoulder)):
                continue
            considered += 1
            # y grows downward, so wrist below shoulder => wrist.y larger.
            drop = (wrist[1] - shoulder[1]) / scale
            if drop > cfg.drop_margin:
                down += 1
                if drop > worst_drop:
                    worst_drop = drop
                    worst_ms = frame.timestamp_ms
        if considered == 0:
            return None, None
        return down / considered, worst_ms

File: app/lib/analysis/rules/hands_up.dart
import '../context.dart';
import '../geometry.dart' as geo;
import '../landmarks.dart';
import '../round_analysis.dart';
import '../rule.dart';

/// Rule: are the hands kept up when *not* punching? Mirror of
/// `src/boxing_coach/analysis/rules/hands_up.py`.
///
/// Distinct from guard return (the moment after a punch). This is the baseline
/// habit: between exchanges, do the wrists stay up near the head, or drift below
/// the shoulder line? We measure the fraction of idle frames spent hands-down.
class HandsUpConfig {
  const HandsUpConfig({
    this.dropMargin = 0.10,
    this.maxDownFraction = 0.25,
    this.checkLead = true,
    this.checkRear = true,
    this.relativeToBaseline = false,
    this.baselineWindowMs = 1000.0,
    this.excuseWhileMoving = false,
    this.movingSpeed = 0.6,
  });

  final double dropMargin;
  final double maxDownFraction;
  final bool checkLead;
  final bool checkRear;
  final bool relativeToBaseline;
  final double baselineWindowMs;
  final bool excuseWhileMoving;
  final double movingSpeed;

  HandsUpConfig copyWith({
    bool? relativeToBaseline,
    bool? excuseWhileMoving,
  }) => HandsUpConfig(
    dropMargin: dropMargin,
    maxDownFraction: maxDownFraction,
    checkLead: checkLead,
    checkRear: checkRear,
    relativeToBaseline: relativeToBaseline ?? this.relativeToBaseline,
    baselineWindowMs: baselineWindowMs,
    excuseWhileMoving: excuseWhileMoving ?? this.excuseWhileMoving,
    movingSpeed: movingSpeed,
  );
}

class HandsUpRule extends Rule {
  HandsUpRule([HandsUpConfig? config]) : _cfg = config ?? const HandsUpConfig();

  final HandsUpConfig _cfg;

  @override
  String get id => 'hands_up';

  @override
  List<Observation> evaluate(AnalysisContext context) {
    final cfg = context.styleProfile.configFor(id, _cfg);
    final idle = _idleFrameIndices(context);
    if (idle.isEmpty) return <Observation>[];

    final stance = context.drill.stance;
    final checked = <Side>{};
    if (cfg.checkLead) checked.add(stance.lead);
    if (cfg.checkRear) checked.add(stance.rear);

    final observations = <Observation>[];
    for (final side in Side.values) {
      if (!checked.contains(side)) continue;
      final (downFraction, worstMs) = _downFraction(context, side, idle, cfg);
      if (downFraction == null) continue;
      if (downFraction > cfg.maxDownFraction) {
        final hand = side == stance.lead ? 'lead' : 'rear';
        observations.add(
          Observation(
            ruleId: id,
            category: SkillCategory.defence,
            severity: downFraction > 0.5 ? Severity.moderate : Severity.minor,
            coachingText:
                'Your $hand hand keeps drifting down between punches. Glue it '
                "to your cheek — you're open to the counter.",
            timestampMs: worstMs,
            metrics: <String, double>{'down_fraction': downFraction},
            highlightLandmarks: <Landmark>[side.wrist],
          ),
        );
      }
    }
    return observations;
  }

  /// Frames that aren't part of any punch (start..end inclusive).
  List<int> _idleFrameIndices(AnalysisContext context) {
    final punching = <int>{};
    for (final p in context.punches) {
      for (var i = p.startIndex; i <= p.endIndex; i++) {
        punching.add(i);
      }
    }
    return <int>[
      for (var i = 0; i < context.sequence.length; i++)
        if (!punching.contains(i)) i,
    ];
  }

  (double?, double?) _downFraction(
    AnalysisContext context,
    Side side,
    List<int> idle,
    HandsUpConfig cfg,
  ) {
    final seq = context.sequence;
    final scale = context.bodyScale;
    final speeds = cfg.excuseWhileMoving ? context.stanceSpeed : null;
    final baseline =
        cfg.relativeToBaseline ? _baselineDrop(context, side, idle, cfg) : 0.0;
    var considered = 0;
    var down = 0;
    double? worstMs;
    var worstDrop = 0.0;
    for (final i in idle) {
      if (speeds != null && speeds[i] > cfg.movingSpeed) continue;
      final frame = seq.frames[i];
      final wrist = geo.framePoint(frame, side.wrist);
      final shoulder = geo.framePoint(frame, side.shoulder);
      if (wrist.any((v) => v.isNaN) || shoulder.any((v) => v.isNaN)) continue;
      considered++;
      final drop = (wrist[1] - shoulder[1]) / scale - baseline;
      if (drop > cfg.dropMargin) {
        down++;
        if (drop > worstDrop) {
          worstDrop = drop;
          worstMs = frame.timestampMs;
        }
      }
    }
    if (considered == 0) return (null, null);
    return (down / considered, worstMs);
  }

  /// Median wrist-below-shoulder over the fighter's settled early carriage — the
  /// reference a low hand is judged against under Soviet (relativeToBaseline).
  double _baselineDrop(
    AnalysisContext context,
    Side side,
    List<int> idle,
    HandsUpConfig cfg,
  ) {
    final seq = context.sequence;
    final scale = context.bodyScale;
    if (idle.isEmpty) return 0.0;
    final t0 = seq.frames[idle.first].timestampMs;
    final early = <double>[];
    final overall = <double>[];
    for (final i in idle) {
      final frame = seq.frames[i];
      final wrist = geo.framePoint(frame, side.wrist);
      final shoulder = geo.framePoint(frame, side.shoulder);
      if (wrist.any((v) => v.isNaN) || shoulder.any((v) => v.isNaN)) continue;
      final drop = (wrist[1] - shoulder[1]) / scale;
      overall.add(drop);
      if (frame.timestampMs - t0 <= cfg.baselineWindowMs) early.add(drop);
    }
    final drops = early.isNotEmpty ? early : overall;
    return drops.isNotEmpty ? geo.nanMedian(drops) : 0.0;
  }
}
